File: python/binarySearch/searchMatrix.py
import sys
import threading
from typing import List, Tuple, Dict, Set
from collections import defaultdict, deque


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        largura = len(matrix[0])
        altura = len(matrix)
        left = 0
        right = (largura * altura) - 1

        while left <= right:
            mid = (left + right) // 2
            i = mid // largura
            j = mid % largura
            if target == matrix[i][j]:
                return True
            elif target > matrix[i][j]:
                left = mid + 1
            else:
                right = mid - 1

        return False


def main():
    s = Solution()
    data = [[1, 2, 4, 8], [10, 11, 12, 13], [14, 20, 30, 40]]
    print(s.searchMatrix(data, 10))


if __name__ == "__main__":
    threading.Thread(target=main).start()

File: cpp/binarySearch/search_matrix.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define pb push_back
#define all(x) x.begin(), x.end()
#define endl '\n'

bool searchMatrix(vector<vector<int>>& matrix, int target) {

    int largura = matrix[0].size();
    int altura = matrix.size();
    int left = 0;
    int right = (largura * altura) - 1;
    while (left <= right) {
        int mid = (right + left) / 2;
        int i = mid / largura;
        int j = mid % largura;
        if (target == matrix[i][j]) return true;
        if (target > matrix[i][j])
            left = mid + 1;
        else
            right = mid - 1;
    }
    return false;
}

int main() {

    ios::sync_with_stdio(0);
    cin.tie(0);

    return 0;
}
